### PR TITLE
Add printTree function to Binary Tree

### PR DESCRIPTION
**Pull request by sevenseasofbri**
- To make modifications in `BooleanTemplate` this can only be done when logic gates are made.
- Will add an `AND` gate to check the workability of the tree structure for the circuit under:
`component > gates > AndGate`

### CODE REVIEW

**Review by hughjazzman — APPROVED** [review]
LGTM, will need JavaDoc and maybe a comment showing the format of the `printTree`?

**Comment by sevenseasofbri** [comment]
> LGTM, will need JavaDoc and maybe a comment showing the format of the `printTree`?

Yup can, I'll do that

**Comment by sevenseasofbri** [comment]
Just realised the name printTree is sort of misleading since the function really just modifies the 'tree' variable and really doesn't print the tree. Will change the name.

File: src/main/java/seedu/duke/component/gates/Gate.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package seedu.duke.component.gates;
+
+public abstract class Gate {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/component/gates/Gate.java:3]: Nice use of `abstract`!

File: src/main/java/seedu/duke/component/gates/Gate.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    protected int output;
+    protected int input;
+
+    public abstract int getOutput();
+
+    public void setInput(byte input) {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/component/gates/Gate.java]: Any reason why the `@param input` here is of `byte` type when everything else is `int`? Including that of `TwoInputGate.setSecondInput()`.

File: src/main/java/seedu/duke/component/gates/AndGate.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    public AndGate(int input1, int input2) {
+        input = input1;
+        secondInput = input2;
+    }
+
+    @Override
+    public int getOutput() {
+        output = input & secondInput;
+        return output;
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/component/gates/AndGate.java]: Probably clear enough to return immediately.
```suggestion
        return input & secondInput;
```

File: src/main/java/seedu/duke/template/BinaryTree.java [hunk 4]
```diff
@@ -11,17 +13,38 @@ public BinaryTree(T root) {
...
     public boolean isEmpty() {
         return arrayList.isEmpty();
     }
 
+    /**
+     * Returns the root of the tree.
+     *
+     * @return T type value, root of the tree.
+     */
     public T getRoot() {
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/template/BinaryTree.java:35]: Should this check for `isEmpty()` first?

File: src/main/java/seedu/duke/template/BinaryTree.java [hunk 5]
```diff
@@ -56,13 +79,40 @@ private boolean isNullAtIndex(int idx) {
...
+    public boolean isLeaf(int index) {
+        if (isNullAtIndex(index) || !isNullAtIndex(getLeftIndex(index)) || !isNullAtIndex(getRightIndex(index))) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Manipulates the instance variable String tree to produce a tree-like structure String output.
+     * Example of manipulated tree String with Integer type values stored in arrayList = {1,2,3,4,5,6,7}:
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/template/BinaryTree.java:96]: Nice example!

File: src/main/java/seedu/duke/template/BinaryTree.java [hunk 6]
```diff
@@ -56,13 +79,40 @@ private boolean isNullAtIndex(int idx) {
...
+     * Returns whether node at specified index is a leaf node or not.
+     *
+     * @param index int type specifying index of node.
+     * @return boolean value, true if node is a leaf, else false.
+     */
+    public boolean isLeaf(int index) {
+        if (isNullAtIndex(index) || !isNullAtIndex(getLeftIndex(index)) || !isNullAtIndex(getRightIndex(index))) {
+            return false;
+        }
+        return true;
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/template/BinaryTree.java]:
```suggestion
        boolean hasNoChildren = isNullAtIndex(getLeftIndex(index)) && isNullAtIndex(getRightIndex(index));
        return !isNullAtIndex(index) && hasNoChildren;
```

File: src/main/java/seedu/duke/template/BinaryTree.java [hunk 7]
```diff
@@ -56,13 +79,40 @@ private boolean isNullAtIndex(int idx) {
...
+     * @param space int type determining number of spaces to make tree-like output.
+     */
+    private void buildTreeString(int index, int space) {
         if (isNullAtIndex(index)) {
             return;
         }
         //Process the right subtree, if it exists.
         if (!isNullAtIndex(getRightIndex(index))) {
-            printTree(getRightIndex(index), space + 4);
+            buildTreeString(getRightIndex(index), space + 4);
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/template/BinaryTree.java]: `4` could be a constant defined above.

**Review by hughjazzman — COMMENTED** [review]
Just some JavaDoc as well I suppose.

File: src/main/java/seedu/duke/component/gates/Gate.java (hunk 2 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/component/gates/Gate.java]: Oops I had initially made it byte then changed everything else to int. Seemed to have missed this one out. Thanks for pointing out!

File: src/main/java/seedu/duke/template/BinaryTree.java (hunk 4 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/template/BinaryTree.java:35]: Actually, the list is initally populated with 16 null values, so technically we gotta check for if the 1st value is null. Might have to change the isEmpty() function.

File: src/main/java/seedu/duke/component/gates/AndGate.java (hunk 3 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/component/gates/AndGate.java]: Noted. So I add set output as a separate fn right?

File: src/main/java/seedu/duke/template/BinaryTree.java (hunk 7 above)
Comment (sevenseasofbri) [reply src/main/java/seedu/duke/template/BinaryTree.java]: Good idea, will do.

File: src/main/java/seedu/duke/component/gates/AndGate.java (hunk 3 above)
Comment (hughjazzman) [reply src/main/java/seedu/duke/component/gates/AndGate.java]: I don't think we need that. Reason being we are not guaranteed the inputs are set when building the circuit. Might have to calculate every time, unless `BooleanTemplate` checks for if a new value is set or not.

Any `Gate` would also need a default overloaded constructor in that case.

Comment (sevenseasofbri) [reply src/main/java/seedu/duke/component/gates/AndGate.java]: Hmm, makes sense

File: src/main/java/seedu/duke/component/gates/AndGate.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+public class AndGate extends TwoInputGate {
+    private static final String AND = "--(&|--";
+
+    public AndGate(int input1, int input2) {
+        input = input1;
+        secondInput = input2;
+    }
+
+    /**
+     * Performs bitwise AND (&) operation on the input instance variables.
```
Comment (hughjazzman) [inline src/main/java/seedu/duke/component/gates/AndGate.java:12]: Also returns it! Found this looking through older PRs and realised this was merged before this review. 😮 

Consider updating the `Gate.getOutput()` methods with better JavaDoc? Goes with #110 I suppose.

**Pull request merged**